src/scram.c

```diff
@@ -862,8 +862,21 @@ char *build_server_first_message(ScramState *scram_state, const char *username,
 				goto failed;
 			break;
 		case PASSWORD_TYPE_PLAINTEXT:
-			if (!build_adhoc_scram_secret(stored_secret, scram_state))
-				goto failed;
+			if (user->adhoc_scram_secrets_cached) {
+				scram_state->iterations = user->iterations;
+				scram_state->salt = strdup(user->salt);
+				memcpy(scram_state->StoredKey, user->StoredKey, sizeof(user->StoredKey));
+				memcpy(scram_state->ServerKey, user->ServerKey, sizeof(user->ServerKey));
```

Comment:
I think we should also do this caching for the `PASSWORD_TYPE_SCRAM_SHA_256` case. So let's move this before the case. So in pseudo code:

```
if cached
    scram_state = cached_scram_state
else
    case PASSWORD_TYPE_SCRAM_SHA_256
	scram_state = parse_scram_secret()
    case PASSWORD_TYPE_PLAINTEXT
	scram_state = build_adhoc_scram_secret()
    cached_scram_state = scram_state
```